node/subsystem-util/src/rolling_session_window.rs

```diff
@@ -94,55 +101,176 @@ pub enum SessionWindowUpdate {
+					stored_window.session_info.clear();
+				}
+
+				// The session window might extend beyond the last finalized block, but that's fine as we'll prune it at
+				// next update.
+				let window_start = if stored_window.session_info.len() > 0 {
+					// If there is at least one entry in db, we always take the DB as source of truth.
+					stored_window.earliest_session
+				} else {
+					on_chain_window_start
+				};
+
+				(window_start, stored_window.session_info)
+			} else {
+				(on_chain_window_start, Vec::new())
+			};
+
+		// Compute the amount of sessions missing from the window that will be fetched from chain state.
+		let sessions_missing_count = session_index
+			.saturating_sub(window_start)
+			.saturating_add(1)
+			.saturating_sub(stored_sessions.len() as u32);
+
+		// Extend from chain state.
+		let sessions = if sessions_missing_count > 0 {
+			match extend_sessions_from_chain_state(
+				stored_sessions,
+				&mut sender,
+				block_hash,
+				&mut window_start,
+				session_index,
+			)
+			.await
+			{
+				Err(kind) => Err(SessionsUnavailable {
+					kind,
+					info: Some(SessionsUnavailableInfo {
+						window_start,
+						window_end: session_index,
+						block_hash,
+					}),
 				}),
-			}),
-			Ok(s) => Ok(Self { earliest_session: window_start, session_info: s, window_size }),
+				Ok(sessions) => Ok(sessions),
+			}?
+		} else {
+			// There are no new sessions to be fetched from chain state.
+			Vec::new()
+		};
+
+		Ok(Self {
+			earliest_session: window_start,
+			session_info: sessions,
+			window_size: SESSION_WINDOW_SIZE,
+			db_params: Some(db_params),
+		})
+	}
+
+	// Load session information from the parachains db.
+	fn db_load(db_params: DatabaseParams) -> Option<StoredWindow> {
+		match db_params.db.get(db_params.db_column, STORED_ROLLING_SESSION_WINDOW).ok()? {
+			None => None,
+			Some(raw) => {
+				let maybe_decoded = StoredWindow::decode(&mut &raw[..]).map(Some);
+				match maybe_decoded {
+					Ok(decoded) => decoded,
+					Err(err) => {
+						gum::warn!(
+							target: LOG_TARGET,
+							?err,
+							"Failed decoding db entry; will start with onchain session infos and self-heal DB entry on next update."
+						);
+						None
+					},
+				}
+			},
+		}
+	}
+
+	// Saves/Updates all sessions in the database.
+	// TODO: https://github.com/paritytech/polkadot/issues/6144
+	fn db_save(&mut self, stored_window: StoredWindow) {
+		if let Some(db_params) = self.db_params.as_ref() {
+			match db_params.db.write(DBTransaction {
+				ops: vec![DBOp::Insert {
+					col: db_params.db_column,
+					key: DBKey::from_slice(STORED_ROLLING_SESSION_WINDOW),
+					value: stored_window.encode(),
+				}],
+			}) {
+				Ok(_) => {},
+				Err(err) => {
+					gum::warn!(target: LOG_TARGET, ?err, "Failed writing db entry");
+				},
+			}
 		}
 	}
 
 	/// Initialize a new session info cache with the given window size and
 	/// initial data.
+	/// This is only used in `approval voting` tests.
```

Comment:
Maybe hide this function under `#[cfg(test)]` then?

Reply:
If I do that, the approval tests don't compile anymore as the function is not visible.